### PR TITLE
3 b training prep

### PR DESCRIPTION
**Pull request by le1nux**
# What does this PR do?

This PR prepares the 3B training path by tightening weight-tying behavior for parallel training and by making DCP checkpoint restores more flexible.

## General Changes
* Add selective `AppState` component loading so checkpoint restore can load only the model, optimizer, and/or LR scheduler as needed.
* Thread `components_to_load` and `allow_partial_load` through the app-state factory and DCP checkpoint loading path.
* Add `has_tied_word_embeddings` model capability checks and centralize tied-embedding validation helpers.
* Reject tied word embeddings for Tensor Parallelism and Pipeline Parallelism configs.
* Update the Llama3-like initializer so `lm_head` is only initialized separately when weight tying is disabled.
* Expose `has_tied_word_embeddings` on GPT-2 models and add a default implementation on the base model class.
* Add tests covering selective checkpoint component loading and tied-embedding validation behavior.

## Breaking Changes
* Tensor Parallelism and Pipeline Parallelism configs now fail validation when tied word embeddings are enabled.
* DCP app-state loading config now includes `allow_partial_load`, which changes how partial checkpoint restores can be configured explicitly.

## Checklist before submitting final PR
- [ ] My PR is minimal and addresses one issue in isolation
- [ ] I have merged the latest version of the target branch into this feature branch
- [ ] I have reviewed my own code w.r.t. correct implementation, missing type hints, proper documentation, etc.
- [ ] I have run a sample config for model training
- [ ] I have checked that all tests run through (`python tests/tests.py`)
- [ ] I have updated the internal changelog (`CHANGELOG_DEV.md`)

### CODE REVIEW

File: src/modalities/checkpointing/stateful/app_state.py [hunk 1]
```diff
@@ -106,12 +120,14 @@ def load_state_dict(self, state_dict: dict[str, Any]) -> None:
...
-        )
-        if self._lr_scheduler is not None:
+        if StatefulComponents.MODEL in self._components_to_load:
+            ModelStateRetriever.load_state_dict_(app_state=self, state_dict=state_dict[StatefulComponents.MODEL.value])
+        if StatefulComponents.OPTIMIZER in self._components_to_load:
+            OptimizerStateRetriever.load_state_dict_(
+                app_state=self,
+                state_dict=state_dict[StatefulComponents.OPTIMIZER.value],
+            )
+        if self._lr_scheduler is not None and StatefulComponents.LR_SCHEDULER in self._components_to_load:
```
Comment (rrutmann) [inline src/modalities/checkpointing/stateful/app_state.py:137]: Should we raise an error if self._components_to_load contains something unexpected?

File: src/modalities/checkpointing/fsdp/fsdp_checkpoint_loading.py [hunk 2]
```diff
@@ -103,17 +103,18 @@ def load_optimizer_checkpoint_(self, optimizer: Optimizer, model: FSDP, file_pat
...
 
-    def __init__(self, global_rank: int):
+    def __init__(self, global_rank: int, allow_partial_load: bool = False):
         """
         Initializes the DCPCheckpointLoading object.
 
         Args:
             global_rank (int): The global rank of the process.
-
+            allow_partial_load (bool, optional): Whether to allow partial loading of the checkpoint. Defaults to True.
```
Comment (rrutmann) [inline src/modalities/checkpointing/fsdp/fsdp_checkpoint_loading.py]:
```suggestion
            allow_partial_load (bool, optional): Whether to allow partial loading of the checkpoint. Defaults to False.
```

File: src/modalities/checkpointing/stateful/app_state_factory.py [hunk 3]
```diff
@@ -25,24 +28,35 @@ def get_raw_app_state(
...
+            lr_scheduler=lr_scheduler,
+            components_to_load=components_to_load,
+        )
         return app_state
 
     @staticmethod
     def get_dcp_checkpointed_app_state_(
         raw_app_state: AppState,
         checkpoint_dir_path: Path,
+        allow_partial_load: bool = True,
```
Comment (rrutmann) [inline src/modalities/checkpointing/stateful/app_state_factory.py]: The default value in DCPCheckpointLoading for allow_partial_load is False. I suggest to make this consistent
```suggestion
        allow_partial_load: bool = False,
```

File: src/modalities/config/config.py [hunk 4]
```diff
@@ -124,8 +126,9 @@ def parse_sharding_strategy_by_name(cls, name: str) -> ShardingStrategy:
         return parse_enum_by_name(name=name, enum_type=ShardingStrategy)
 
 
-class DCPCheckpointLoadingConfig(BaseModel):
-    global_rank: Annotated[int, Field(strict=True, ge=0)]
+# class DCPCheckpointLoadingConfig(BaseModel):
+#     global_rank: Annotated[int, Field(strict=True, ge=0)]
+#     allow_partial_load: bool = True
```
Comment (rrutmann) [inline src/modalities/config/config.py]: Can we remove this?

File: tests/checkpointing/test_app_state_components_to_load.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
+from unittest.mock import MagicMock
```
Comment (rrutmann) [inline tests/checkpointing/test_app_state_components_to_load.py:1]: Maybe add a test for invalid combinations of allow_partial_load and components_to_load

**Review by rrutmann — APPROVED** [review]

File: src/modalities/checkpointing/stateful/app_state.py (hunk 1 above)
Comment (le1nux) [reply src/modalities/checkpointing/stateful/app_state.py:137]: added a check and also a test case for this.

File: tests/checkpointing/test_app_state_components_to_load.py (hunk 5 above)
Comment (le1nux) [reply tests/checkpointing/test_app_state_components_to_load.py:1]: done
